docs/language/ql-training-rst/cpp/bad-overflow-guard.rst

```diff
@@ -0,0 +1,228 @@
+
+Finding bad overflow guards
+
+QL query: bad overflow guards
+=============================
+
+Let’s look for overflow guards of the form ``v + b < v``, using the classes
+``AddExpr``, ``Variable`` and ``RelationalOperation`` from the ``cpp`` library.
+
+.. rst-class:: build
+
+.. literalinclude:: ../query-examples/cpp/bad-overflow-guard-1.ql
+   :language: ql
+
+.. note::
+
+  - When performing `variant analysis <https://semmle.com/variant-analysis>`__, it is usually helpful to write a simple query that finds the simple syntactic pattern, before trying to go on to describe the cases where it goes wrong.
+  - In this case, we start by looking for all the *overflow* checks, before trying to refine the query to find all *bad overflow* checks.
+  - The select clause defines what this query is looking for:
+    - an ``AddExpr``: the expression that is being checked for overflow.
+    - a ``RelationalOperation``: the overflow comparison check.
+    - a ``Variable``: used as an argument to both the addition and comparison.
+  - The where part of the query ties these three QL variables together using `predicates <https://help.semmle.com/QL/ql-handbook/predicates.html>`__ defined in the `standard QL for C/C++ library <https://help.semmle.com/qldoc/cpp/>`__.
+
+QL query: bad overflow guards
+=============================
+
+We want to ensure the operands being added have size less than 4 bytes.
+
+We may want to reuse this logic, so let us create a separate predicate.
+
+Looking at autocomplete suggestions, we see that we can get the type of an expression using the ``getType()`` method.
+
+We can get the size (in bytes) of a type using the ``getSize()`` method.
+
+.. rst-class:: build
+
+.. code-block:: ql
+
+  predicate isSmall(Expr e) {
+    e.getType().getSize() < 4
+  }
+
+.. note::
+
+  - An important part of the query is to determine whether a given expression has a “small” type that is going to trigger integer promotion.
+  - We therefore write a helper predicate for small expressions.
+  - This predicate effectively represents the set of all expressions in the database where the size of the type of the expression is less than 4 bytes, that is less than 32 bits.
+
+QL query: bad overflow guards
+=============================
+
+We can ensure the operands being added have size less than 4 bytes, using our new predicate.
+QL has logical quantifiers like ``exists`` and ``forall``, allowing us to declare variables and enforce a certain condition on them.
+Now our query becomes:
+
+.. rst-class:: build
+
+.. literalinclude:: ../query-examples/cpp/bad-overflow-guard-2.ql
+   :language: ql
+
+.. note::
+
+  - Recall from earlier that what makes an overflow check a “bad” check is that all the arguments to the addition are integers smaller than 32 bits.
+  - We could write this by using our helper predicate ``isSmall`` to specify that each individual operand to the addition ``isSmall`` (that is under 32 bits):
+
+  .. code-block:: ql
```

Comment:
It looks as if you can indent this so that it lines up with the bullets: https://sublime-and-sphinx-guide.readthedocs.io/en/latest/lists.html?#code-images-and-other-content-inside-lists